kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/LogCompactionWriterSuite.scala

```diff
@@ -0,0 +1,281 @@
+import org.apache.spark.sql.delta.actions.{DomainMetadata => DeltaSparkDomainMetadata, TableFeatureProtocolUtils}
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+
+/**
+ * Test suite for io.delta.kernel.internal.compaction.LogCompactionWriter
+ */
+class LogCompactionWriterSuite extends CheckpointSuiteBase {
+  val COMPACTED_SCHEMA =
+    new StructType()
+      .add("txn", SetTransaction.FULL_SCHEMA)
+      .add("add", AddFile.FULL_SCHEMA)
+      .add("remove", RemoveFile.FULL_SCHEMA)
+      .add("metaData", Metadata.FULL_SCHEMA)
+      .add("protocol", Protocol.FULL_SCHEMA)
+      // .add("cdc", new StructType())
+      .add("domainMetadata", DomainMetadata.FULL_SCHEMA);
+
+  // check if a row is all null
+  def rowIsNull(row: Row): Boolean = {
+    val schema = row.getSchema()
+    for (ordinal <- 0 until schema.length()) {
+      if (!row.isNullAt(ordinal)) {
+        return false
+      }
+    }
+    true
+  }
+
+  // Get the expected actions from the log.  We filter down to just the actions that end up in a
+  // compacted log file, and also filter out adds that have been removed as well as duplicate
+  // metadata/protocol actions
+  def getActionsFromLog(
+      tablePath: Path,
+      engine: Engine,
+      startVersion: Long,
+      endVersion: Long): Seq[TestRow] = {
+    val files = DeltaLogActionUtils.listDeltaLogFilesAsIter(
+      engine,
+      new HashSet(Arrays.asList(DeltaLogFileType.COMMIT)),
+      tablePath,
+      startVersion,
+      Optional.of(endVersion),
+      false /* mustBeRecreatable */ )
+      .toInMemoryList()
+    Collections.reverse(files) // we want things in reverse order
+    val actions =
+      new ActionsIterator(engine, files, COMPACTED_SCHEMA, Optional.empty())
+    val removed = scala.collection.mutable.HashSet.empty[String]
+    var seenMetadata = false
+    var seenProtocol = false
+    val resBuilder = Seq.newBuilder[TestRow]
+    while (actions.hasNext()) {
+      val batch = actions.next().getColumnarBatch()
+      val rows = batch.getRows()
+      while (rows.hasNext()) {
+        val row = rows.next()
+
+        if (!row.isNullAt(2)) {
+          // remove
+          val removeRow = row.getStruct(2)
+          val path = removeRow.getString(0)
+          removed += path
+        }
+
+        if (!row.isNullAt(1)) {
+          // add
+          val addRow = row.getStruct(1)
+          val path = addRow.getString(0)
+          if (!removed.contains(path)) {
+            resBuilder += TestRow(row)
+          }
+        } else if (!row.isNullAt(3)) {
+          // metadata
+          if (!seenMetadata) {
+            seenMetadata = true
+            resBuilder += TestRow(row)
+          }
+        } else if (!row.isNullAt(4)) {
+          // protocol
+          if (!seenProtocol) {
+            seenProtocol = true
+            resBuilder += TestRow(row)
+          }
+        } else if (!rowIsNull(row)) {
+          resBuilder += TestRow(row)
```

Comment:
what do we do with domain metadatas here? don't those also have to be resolved with themselves? (only include the latest one)

Reply:
Great catch, we need to resolve those too. I've added that into the test now